src/bp/core/services/messaging/messaging.ts

```diff
@@ -0,0 +1,268 @@
+import * as sdk from 'botpress/sdk'
+import { ConversationRepository } from 'core/repositories/conversations'
+import { MessageRepository } from 'core/repositories/messages'
+import { IOEvent } from 'core/sdk/impl'
+import { inject, injectable, postConstruct } from 'inversify'
+import { AppLifecycle, AppLifecycleEvents } from 'lifecycle'
+import LRU from 'lru-cache'
+import ms from 'ms'
+import { TYPES } from '../../types'
+import { JobService } from '../job-service'
+import { KeyValueStore } from '../kvs'
+import { EventEngine } from '../middleware/event-engine'
+
+@injectable()
+export class MessagingAPI {
+  private mostRecentCaches: { [botId: string]: LRU<string, sdk.experimental.Conversation> } = {}
+  private lastChannelCaches: { [botId: string]: LRU<string, string> } = {}
+  private invalidateMostRecent: (endpoint: sdk.experimental.UserEndpoint, mostRecentConvoId?: number) => void = this
+    ._localInvalidateMostRecent
+  private invalidateLastChannel: (endpoint: sdk.experimental.UserEndpoint, lastChannel: string) => void = this
+    ._localInvalidateLastChannel
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.JobService) private jobService: JobService,
+    @inject(TYPES.MessageRepository) private messageRepo: MessageRepository,
+    @inject(TYPES.ConversationRepository) private conversationRepo: ConversationRepository,
+    @inject(TYPES.KeyValueStore) private kvs: KeyValueStore
+  ) {}
+
+  @postConstruct()
+  async init() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.CONFIGURATION_LOADED)
+
+    this.invalidateMostRecent = <any>await this.jobService.broadcast<void>(this._localInvalidateMostRecent.bind(this))
+    this.invalidateLastChannel = <any>await this.jobService.broadcast<void>(this._localInvalidateLastChannel.bind(this))
+
+    this.eventEngine.onSendIncoming = async event => {
+      await this.updateLastChannel({ userId: event.target, botId: event.botId }, event.channel)
+    }
+  }
+
+  public async getAllConversations(endpoint: sdk.experimental.UserEndpoint): Promise<sdk.experimental.Conversation[]> {
+    return this.conversationRepo.getAll(endpoint)
+  }
+
+  public async getRecentConversations(
+    endpoint: sdk.experimental.UserEndpoint,
+    limit?: number
+  ): Promise<sdk.experimental.RecentConversation[]> {
+    return this.conversationRepo.getAllRecent(endpoint, limit)
+  }
+
+  public async deleteAllConversations(endpoint: sdk.experimental.UserEndpoint): Promise<number> {
+    this.invalidateMostRecent(endpoint)
+
+    return this.conversationRepo.deleteAll(endpoint)
+  }
+
+  public async createConversation(endpoint: sdk.experimental.UserEndpoint): Promise<sdk.experimental.Conversation> {
+    return this.conversationRepo.create(endpoint)
+  }
+
+  public async getOrCreateRecentConversation(
```

Comment:
why not `getOrCreateConversation`  XD  
I mean this method will work even if the conversation isn't recent, correct ?